jina/clients/base/unary_rpc.py

```diff
@@ -0,0 +1,126 @@
+import asyncio
+from typing import TYPE_CHECKING, Optional, Tuple
+
+import grpc
+
+from jina.clients.base.retry import wait_or_raise_err
+from jina.clients.helper import callback_exec
+from jina.excepts import InternalNetworkError
+from jina.proto import jina_pb2_grpc
+from jina.serve.stream import RequestStreamer
+
+if TYPE_CHECKING:
+    from jina.types.request import Request
+
+
+class UnaryRpc:
+    """Class that encapsulated the methods required to run unary rpc calls from the client. Instantiate a single class
+    for each client request.
+    """
+
+    def __init__(
+        self,
+        channel,
+        continue_on_error,
+        metadata,
+        on_always,
+        on_done,
+        on_error,
+        p_bar,
+        req_iter,
+        max_attempts,
+        backoff_multiplier,
+        initial_backoff,
+        max_backoff,
+        logger,
+        show_progress,
+        compression,
+        client_args,
+        prefetch,
+        results_in_order,
+        **kwargs
+    ):
+        self.results_in_order = results_in_order
+        self.prefetch = prefetch
+        self.client_args = client_args
+        self.compression = compression
+        self.show_progress = show_progress
+        self.logger = logger
+        self.max_backoff = max_backoff
+        self.initial_backoff = initial_backoff
+        self.backoff_multiplier = backoff_multiplier
+        self.max_attempts = max_attempts
+        self.req_iter = req_iter
+        self.p_bar = p_bar
+        self.on_error = on_error
+        self.on_done = on_done
+        self.on_always = on_always
+        self.metadata = metadata
+        self.continue_on_error = continue_on_error
+        self.channel = channel
+        self.kwargs = kwargs
+
+    async def unary_rpc_with_retry(self):
+        """Wraps the unary rpc call with retry loop based on the retry params.
+        :yields: Responses received from the target.
+        """
+        stub = jina_pb2_grpc.JinaSingleDataRequestRPCStub(self.channel)
+
+        def _request_handler(
+            request: 'Request',
+        ) -> 'Tuple[asyncio.Future, Optional[asyncio.Future]]':
+            async def _with_retry(req: 'Request'):
```

Comment:
where `_with_retry` is imported? Is it an error here?

Reply:
ok, i see, the `_with_retry` is implemented in a mixin.

Reply:
hold on, I'm a bit confusing. It seems `_with_retry` implementation (`jina/clients/mixin.py`) supports multiple attempts. Why do we use a for-loop below to retry `max_attempts` .

Reply:
I need to update the code but good catch. The `_with_retry` in the (jina/clients/minxin.py) must only be used when sync client and when `stream=True`. If `stream=False` then the retry here must take effect. There are a lot of considerations and optimizations to make it easy for the user.